### PR TITLE
SoulboundCore

### LINKED ISSUE #16: `SoulboundCore.sol` Template
This template will have the following extensions

- Soulbound EIP
- Ownable
- With Signature to enable revoke and issue

### PR DESCRIPTION
**Pull request by 0xfps**
Fixes #16 

## Implemented the SoulboundCore template.

Added the SoulboundCore template, required according to issue #16 .

Environment

| Environment                                       | Version           |
| ------------------------------------------------- | ----------------- |
| Compiler version                                  |   0.8.7    |
| Target EVM version (as per compiler settings):    |  default   |
| Framework/IDE (e.g. Truffle or Remix)             |  remix     |
| Operating system (macOS / Windows)                |  Windows   |

### Description

Includes modification of the SoulboundCore contract.

### CODE REVIEW

File: src/extensions/SoulboundWithSignature.sol [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import "./SoulboundCore.sol";
+
+/**
+* @title SoulboundWithSignature.
+* @author Anthony (fps) https://github.com/fps8k.
+* @dev 
+*
+*/
+contract SoulboundWithSignature is SoulboundCore {
+    uint256 internal immutable CAP;
```
Comment (koolamusic) [inline src/extensions/SoulboundWithSignature.sol]: variables should be `supply` and `MAX_SUPPLY`

File: src/extensions/SoulboundCore.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import "./Soulbound.sol";
+import "./Ownable.sol";
+import "./Allowlist.sol";
+
+/**
+* @title Soulbound Core Contract.
+* @author Daccred.
+* @dev Soulbound Core template.
+*/
+contract SoulboundCore is Ownable, Soulbound, Allowlist {
```
Comment (koolamusic) [inline src/extensions/SoulboundCore.sol:20]: The supply and max supply implementation should actually be a part of Core.

**Review by koolamusic — CHANGES_REQUESTED** [review]

File: src/contracts/interfaces/ISoulbound.sol [hunk 3]
```diff
@@ -27,31 +25,7 @@ import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
...
-    function symbol() external view returns (string memory);
-
-    /**
-     * @dev Returns the Uniform Resource Identifier (URI) for `tokenId` token.
-     */
-    function tokenURI(uint256 tokenId) external view returns (string memory);
-}
-
-
-interface ISoulbond is IERC165, IERC721Metadata {
```
Comment (koolamusic) [inline src/contracts/interfaces/ISoulbound.sol:54]: Soulbound should implement ERC721 Metadata interface. so we keep this

File: src/extensions/Soulbound.sol [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    constructor(string memory name, string memory symbol) ERC4973(name, symbol) {}
+    
+    /// @dev Emitted when a token is minted to an address.
+    event MintSoulboundToken(
+        address indexed from, 
+        address indexed to, 
+        uint256 indexed tokenId
+    );
+    /// @dev Emitted when a token is burnt.
+    event BurnSoulboundToken(
```
Comment (koolamusic) [inline src/extensions/Soulbound.sol]: The events should be 

```
  /// @dev This emits when a new token is created and bound to an account by
  /// any mechanism.
  /// Note: For a reliable `from` parameter, retrieve the transaction's
  /// authenticated `from` field.
  event Attest(address indexed to, uint256 indexed tokenId);
  /// @dev This emits when an existing ABT is revoked from an account and
  /// destroyed by any mechanism.
  /// Note: For a reliable `from` parameter, retrieve the transaction's
  /// authenticated `from` field.
  event Revoke(address indexed to, uint256 indexed tokenId);

```

Comment (0xfps) [reply src/extensions/Soulbound.sol]: The events are emitted from the ERC4973 contract when a token is minted, which the Soulbound inherits.

Comment (koolamusic) [reply src/extensions/Soulbound.sol]: yep. We just need the name of the event to be the same with the spec.

File: src/extensions/SoulboundCore.sol (hunk 2 above)
Comment (0xfps) [reply src/extensions/SoulboundCore.sol:20]: Done

Comment (0xfps) [reply src/extensions/SoulboundCore.sol:20]: Done

**Pull request merged**